### PR TITLE
feat: add `handleSynPackage` to `StakeHub`

### PR DESCRIPTION
**Pull request by Pythonberg1997**
### Description

This pr is to add `handleSynPackage` to `StakeHub` to support cross-chain migration.

### Changes

Notable changes:
* add `handleSynPackage` to `StakeHub`

### CODE REVIEW

File: contracts/BC_fusion/System.sol [hunk 1]
```diff
@@ -3,6 +3,9 @@ pragma solidity 0.8.17;
 
 contract System {
     /*----------------- constants -----------------*/
+    uint8 public constant STAKING_CHANNELID = 0x08;
+    uint8 public constant BC_FUSION_CHANNELID = 0x11;
```
Comment (zjubfd) [inline contracts/BC_fusion/System.sol:7]: Please comment what is the usage of this channel, as `BC_FUSION_CHANNEL` can not explain it.

File: contracts/BC_fusion/StakeHub.sol [hunk 2]
```diff
@@ -137,6 +148,14 @@ contract StakeHub is System, Initializable {
     mapping(address => bool) public blackList;
 
     /*----------------- structs and events -----------------*/
+    struct MigrationPackage {
+        address operatorAddress;
+        address delegator;
+        address refundAddress;
+        uint256 amount;
+        bool delegateVotePower;
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:156]: I think we don't need `delegateVotePower ` in this case.

File: contracts/BC_fusion/StakeHub.sol [hunk 3]
```diff
@@ -137,6 +148,14 @@ contract StakeHub is System, Initializable {
     mapping(address => bool) public blackList;
 
     /*----------------- structs and events -----------------*/
+    struct MigrationPackage {
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:151]: ==> `StakeMigrationPackage`

File: contracts/BC_fusion/StakeHub.sol [hunk 4]
```diff
@@ -137,6 +148,14 @@ contract StakeHub is System, Initializable {
...
+        address operatorAddress;
+        address delegator;
+        address refundAddress;
+        uint256 amount;
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:155]: Please add comments for all fields. Also the decimal of the amount here.

File: contracts/BC_fusion/StakeHub.sol [hunk 5]
```diff
@@ -201,6 +220,11 @@ contract StakeHub is System, Initializable {
     event Claimed(address indexed operatorAddress, address indexed delegator, uint256 bnbAmount);
     event Paused();
     event Resumed();
+    event MigrationSucceed(
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:223]: ==> `MigrateSuccess`

File: contracts/BC_fusion/StakeHub.sol [hunk 6]
```diff
@@ -925,6 +985,59 @@ contract StakeHub is System, Initializable {
...
+                break;
+            }
+            ++idx;
+        }
+
+        return (migrationPackage, success);
+    }
+
+    function _doMigration(MigrationPackage memory migrationPkg) internal returns (uint8) {
+        if (!_validatorSet.contains(migrationPkg.operatorAddress)) {
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:1016]: `delegator` can not be empty address

File: contracts/BC_fusion/StakeHub.sol [hunk 7]
```diff
@@ -925,6 +985,59 @@ contract StakeHub is System, Initializable {
...
+        }
+
+        uint256 shares =
+            IStakeCredit(valInfo.creditContract).delegate{ value: migrationPkg.amount }(migrationPkg.delegator);
+        emit Delegated(migrationPkg.operatorAddress, migrationPkg.delegator, shares, migrationPkg.amount);
+        emit MigrationSucceed(migrationPkg.operatorAddress, migrationPkg.delegator, shares, migrationPkg.amount);
+
+        IGovToken(GOV_TOKEN_ADDR).sync(valInfo.creditContract, migrationPkg.delegator);
+        if (migrationPkg.delegateVotePower) {
+            IGovToken(GOV_TOKEN_ADDR).delegateVote(migrationPkg.delegator, migrationPkg.operatorAddress);
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:1035]: No need for this.

File: contracts/BC_fusion/StakeHub.sol [hunk 8]
```diff
@@ -11,13 +11,23 @@ import "./interface/IBSCValidatorSet.sol";
...
 
 contract StakeHub is System, Initializable {
+    using RLPDecode for *;
     using Utils for string;
     using Utils for bytes;
     using EnumerableSet for EnumerableSet.AddressSet;
 
     /*----------------- constants -----------------*/
+    // migration res code
+    uint8 public constant MIGRATION_SUCCESS = 0;
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:25]: Good to use Enum?

File: contracts/BC_fusion/lib/RLPDecode.sol [hunk 9]
```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.17;
+
```
Comment (zjubfd) [inline contracts/BC_fusion/lib/RLPDecode.sol:3]: Can you comment  where is file comes from, and this is unnecessary to be auditted.

File: contracts/BC_fusion/StakeHub.sol (hunk 3 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:151]: Done

File: contracts/BC_fusion/StakeHub.sol (hunk 4 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:155]: Done

File: contracts/BC_fusion/StakeHub.sol (hunk 5 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:223]: Done

File: contracts/BC_fusion/StakeHub.sol [hunk 10]
```diff
@@ -201,6 +220,11 @@ contract StakeHub is System, Initializable {
     event Claimed(address indexed operatorAddress, address indexed delegator, uint256 bnbAmount);
     event Paused();
     event Resumed();
+    event MigrationSucceed(
+        address indexed operatorAddress, address indexed delegator, uint256 shares, uint256 bnbAmount
+    );
+    event MigrationFailed(address indexed operatorAddress, address indexed delegator, uint256 bnbAmount, uint8 resCode);
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:226]: Done

File: contracts/BC_fusion/StakeHub.sol [hunk 11]
```diff
@@ -925,6 +985,59 @@ contract StakeHub is System, Initializable {
...
+            } else {
+                break;
+            }
+            ++idx;
+        }
+
+        return (migrationPackage, success);
+    }
+
+    function _doMigration(MigrationPackage memory migrationPkg) internal returns (uint8) {
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:1015]: Done

File: contracts/BC_fusion/StakeHub.sol (hunk 6 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:1016]: Done

File: contracts/BC_fusion/StakeHub.sol (hunk 8 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:25]: Done

File: contracts/BC_fusion/lib/RLPDecode.sol (hunk 9 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/lib/RLPDecode.sol:10]: Done

File: contracts/BC_fusion/System.sol (hunk 1 above)
Comment (zjubfd) [reply contracts/BC_fusion/System.sol]: `new channel id for cross-chain stake migration after BC-fusion` ==> `new channel id for cross-chain redelegate from Beacon Chain to Smart Chain after Feyman upgrade`

File: contracts/System.sol [hunk 12]
```diff
@@ -18,6 +18,7 @@ contract System {
   uint8 constant public GOV_CHANNELID = 0x09;
   uint8 constant public SLASH_CHANNELID = 0x0b;
   uint8 constant public CROSS_STAKE_CHANNELID = 0x10;
+  uint8 constant public BC_FUSION_CHANNELID = 0x11; // new channel id for cross-chain stake migration after BC-fusion
```
Comment (cosinlink) [inline contracts/System.sol:21]: where do it register `BC_FUSION_CHANNELID ` ?

File: contracts/BC_fusion/StakeHub.sol [hunk 13]
```diff
@@ -258,6 +283,45 @@ contract StakeHub is System, Initializable {
...
 
+    /*----------------- Implement cross chain app -----------------*/
+    function handleSynPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract returns (bytes memory) {
+        (StakeMigrationPackage memory migrationPkg, bool decodeSuccess) = _decodeMigrationSynPackage(msgBytes);
+        if (!decodeSuccess) revert InvalidSynPackage();
+
+        // claim fund from TokenHub
+        _isReceivingFund = 1;
+        bool claimSuccess = ITokenHub(TOKEN_HUB_ADDR).claimMigrationFund(migrationPkg.amount);
+        if (!claimSuccess) {
```
[the remark this answers is not in the document]
Comment (zjubfd) [reply contracts/BC_fusion/StakeHub.sol:294]: Good point

File: contracts/BC_fusion/StakeHub.sol [hunk 14]
```diff
@@ -137,6 +141,22 @@ contract StakeHub is System, Initializable {
     mapping(address => bool) public blackList;
 
     /*----------------- structs and events -----------------*/
+    struct StakeMigrationPackage {
+        address operatorAddress; // the operator address of the target validator to delegate to
+        address delegator; // the beneficiary of the delegation
+        address refundAddress; // the BC address to refund the fund if migration failed
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:147]: `BC` ==> `Beacon Chain`

File: contracts/BC_fusion/StakeHub.sol [hunk 15]
```diff
@@ -137,6 +141,22 @@ contract StakeHub is System, Initializable {
...
+        uint256 amount; // the amount of BNB to be migrated(decimal: 18)
+    }
+
+    enum StakeMigrationStatus {
```
Comment (zjubfd) [inline contracts/BC_fusion/StakeHub.sol:151]: => `StakeMigrationRespCode`

File: contracts/TokenHub.sol [hunk 16]
```diff
@@ -167,6 +167,14 @@ contract TokenHub is ITokenHub, System, IParamSubscriber, IApplication, ISystemR
     return actualAmount;
   }
 
+  function claimMigrationFund(uint256 amount) onlyStakeHub external returns(bool) {
+    if (address(this).balance >= amount) {
+      payable(STAKE_HUB_ADDR).transfer(amount);
```
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply contracts/TokenHub.sol:172]: Done. Added to `StakeHub`

File: contracts/BC_fusion/StakeHub.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:294]: Done

File: contracts/System.sol (hunk 12 above)
Comment (Pythonberg1997) [reply contracts/System.sol:21]: By governance

File: contracts/BC_fusion/StakeHub.sol (hunk 14 above)
Comment (Pythonberg1997) [reply contracts/BC_fusion/StakeHub.sol:147]: Done

**Review by zjubfd — APPROVED** [review]

**Review by cosinlink — APPROVED** [review]

**Pull request merged**